File: structural/typescript/proxy_structured/protection_proxy.ts
export interface Database {
  query(sql: string): void;
}

export class RealDatabase implements Database {
  query(sql: string): void {
    console.log(`Executing query: ${sql}`);
  }
}

export class DatabaseProxy implements Database {
  private realDatabase?: RealDatabase;
  private currentUser: string;
  private permissions: Map<string, boolean> = new Map();
  
  constructor(user: string) {
    this.currentUser = user;
    this.permissions.set('admin', true);
    this.permissions.set('user', false);
    this.permissions.set('guest', false);
  }
  
  private hasPermission(): boolean {
    return this.permissions.get(this.currentUser) ?? false;
  }
  
  query(sql: string): void {
    if (this.hasPermission()) {
      if (!this.realDatabase) {
        this.realDatabase = new RealDatabase();
      }
      this.realDatabase.query(sql);
    } else {
      console.log(`Access denied: User '${this.currentUser}' does not have permission to execute queries`);
    }
  }
  
  setUser(user: string): void {
    this.currentUser = user;
  }
}

File: structural/dart/proxy_structured/protection_proxy.dart
abstract class Database {
  void query(String sql);
}

class RealDatabase extends Database {
  @override
  void query(String sql) {
    print('Executing query: $sql');
  }
}

class DatabaseProxy extends Database {
  RealDatabase? _realDatabase;
  String _currentUser;
  final Map<String, bool> _permissions = {};
  
  DatabaseProxy(this._currentUser) {
    _permissions['admin'] = true;
    _permissions['user'] = false;
    _permissions['guest'] = false;
  }
  
  bool _hasPermission() {
    return _permissions[_currentUser] ?? false;
  }
  
  @override
  void query(String sql) {
    if (_hasPermission()) {
      _realDatabase ??= RealDatabase();
      _realDatabase!.query(sql);
    } else {
      print('Access denied: User \'$_currentUser\' does not have permission to execute queries');
    }
  }
  
  void setUser(String user) {
    _currentUser = user;
  }
}
